### ISSUE TITLE
Combining boolean predicates

### ISSUE THREAD

**Issue by matwojcik** [fthomas/refined#917]
I have some problem when playing with refines using boolean predicates. 

```scala
  object test {
    import eu.timepit.refined.auto._
    type LongOrDouble = NonEmpty And (ValidLong Or ValidDouble)

    val SomeLong: String Refined (NonEmpty And ValidLong) = "11"
    val SomeLongOrDouble: String Refined LongOrDouble = SomeLong


  }
```

which fails with

```
[error] type mismatch;
[error]  found   : eu.timepit.refined.api.Refined[String,eu.timepit.refined.boolean.And[eu.timepit.refined.collection.NonEmpty,eu.timepit.refined.string.ValidLong]]
[error]     (which expands to)  eu.timepit.refined.api.Refined[String,eu.timepit.refined.boolean.And[eu.timepit.refined.boolean.Not[eu.timepit.refined.collection.Empty],eu.timepit.refined.string.ValidLong]]
[error]  required: eu.timepit.refined.api.Refined[String,test.LongOrDouble]
[error]     (which expands to)  eu.timepit.refined.api.Refined[String,eu.timepit.refined.boolean.And[eu.timepit.refined.boolean.Not[eu.timepit.refined.collection.Empty],eu.timepit.refined.boolean.Or[eu.timepit.refined.string.ValidLong,eu.timepit.refined.string.ValidDouble]]]
[error]   val SomeLongOrDouble: String Refined LongOrDouble = SomeLong
[error]                                                       ^

```

```scala
object test2 {
  import eu.timepit.refined.auto._
  type LongOrDouble = (NonEmpty And ValidLong) Or (NonEmpty And ValidDouble)

  val SomeLongOrDouble: String Refined LongOrDouble = "11"

  val SomeNonEmpty: String Refined NonEmpty = SomeLongOrDouble
}

```

which fails with

```
[error] type mismatch;
[error]  found   : eu.timepit.refined.api.Refined[String,test2.LongOrDouble]
[error]     (which expands to)  eu.timepit.refined.api.Refined[String,eu.timepit.refined.boolean.Or[eu.timepit.refined.boolean.And[eu.timepit.refined.boolean.Not[eu.timepit.refined.collection.Empty],eu.timepit.refined.string.ValidLong],eu.timepit.refined.boolean.And[eu.timepit.refined.boolean.Not[eu.timepit.refined.collection.Empty],eu.timepit.refined.string.ValidDouble]]]
[error]  required: eu.timepit.refined.api.Refined[String,eu.timepit.refined.collection.NonEmpty]
[error]     (which expands to)  eu.timepit.refined.api.Refined[String,eu.timepit.refined.boolean.Not[eu.timepit.refined.collection.Empty]]
[error]   val SomeNonEmpty: String Refined NonEmpty = SomeLongOrDouble
[error]                                  
```

I'm not sure if this is limitation of scala as a language or is there some possibility for that to work?

**Comment by matwojcik** [comment]
This does not resolve the issue I described, but pretty much resolves the initial issue I had - is there any way to imply that if a `String` is a `ValidLong` then it must be `NonEmpty` by the long definition?

**Comment by kubukoz** [comment]
Looks like an `Inference` (https://github.com/fthomas/refined/blob/b9b027b4f3a42409b656a2331bfccdff1c922751/modules/core/shared/src/main/scala/eu/timepit/refined/api/Inference.scala) instance is missing for `A and B` ==> `A` / `A and B` ==> `B` 🤔

**Comment by matwojcik** [comment]
@kubukoz nice! Is this mechanism also a good fit for `ValidLong ==> NonEmptyString` so my original issue, which lead me to this?

**Comment by kubukoz** [comment]
I think so. :)

**Comment by fthomas** [comment]
Fixed by #927.
